File: macos/Sources/FloatingLyricCore/Playback/PlaybackController.swift
import Foundation

/// Play/pause and track skipping through the Spotify Web API.
///
/// Every one of these needs the `user-modify-playback-state` scope and a
/// Premium account — Spotify rejects control calls from free accounts with 403.
public final class PlaybackController: @unchecked Sendable {
    private static let base = "https://api.spotify.com/v1/me/player"

    private let auth: SpotifyAuth
    private let http: HTTPClient

    public init(auth: SpotifyAuth, http: HTTPClient) {
        self.auth = auth
        self.http = http
    }

    public func play() async -> AppError? { await send(path: "/play", method: "PUT") }
    public func pause() async -> AppError? { await send(path: "/pause", method: "PUT") }
    public func next() async -> AppError? { await send(path: "/next", method: "POST") }
    public func previous() async -> AppError? { await send(path: "/previous", method: "POST") }

    public func toggle(isPlaying: Bool) async -> AppError? {
        isPlaying ? await pause() : await play()
    }

    /// Returns nil on success, or the error to show in the panel.
    private func send(path: String, method: String) async -> AppError? {
        do {
            let response = try await request(path: path, method: method)
            if response.status == 401 {
                auth.invalidateAccessToken()
                let retry = try await request(path: path, method: method)
                return classify(retry)
            }
            return classify(response)
        } catch let error as AppError {
            return error
        } catch {
            return .network(error.localizedDescription)
        }
    }

    private func request(path: String, method: String) async throws -> HTTPResponse {
        var request = URLRequest(url: URL(string: Self.base + path)!)
        request.httpMethod = method
        request.setValue("Bearer \(try await auth.accessToken())",
                         forHTTPHeaderField: "Authorization")
        // Spotify rejects a PUT with no body and no length header.
        request.setValue("0", forHTTPHeaderField: "Content-Length")
        return try await http.send(request)
    }

    private func classify(_ response: HTTPResponse) -> AppError? {
        switch response.status {
        case 200..<300:
            return nil
        case 401:
            return .sessionExpired
        case 403:
            // 403 covers both "free account" and "token predates the control
            // scope"; the body distinguishes them.
            let body = String(decoding: response.body, as: UTF8.self).lowercased()
            if body.contains("scope") { return .controlScopeMissing }
            return .premiumRequired
        case 404:
            return .noActiveDevice
        case 429:
            let retryAfter = TimeInterval(response.headers["Retry-After"] ?? "") ?? 5
            return .rateLimited(retryAfter: retryAfter)
        default:
            return .badResponse(status: response.status)
        }
    }
}

File: windows/src/FloatingLyric.Core/PlaybackController.cs
namespace FloatingLyric.Core;

/// <summary>
/// Play, pause and skip. Every one of these needs the
/// user-modify-playback-state scope and a Premium account — Spotify rejects
/// control calls from free accounts with 403.
/// </summary>
public sealed class PlaybackController(SpotifyAuth auth, IHttpClient http)
{
    private const string Base = "https://api.spotify.com/v1/me/player";

    public Task<AppError?> PlayAsync() => SendAsync("/play", HttpMethod.Put);
    public Task<AppError?> PauseAsync() => SendAsync("/pause", HttpMethod.Put);
    public Task<AppError?> NextAsync() => SendAsync("/next", HttpMethod.Post);
    public Task<AppError?> PreviousAsync() => SendAsync("/previous", HttpMethod.Post);

    public Task<AppError?> ToggleAsync(bool isPlaying) =>
        isPlaying ? PauseAsync() : PlayAsync();

    /// <summary>Null on success, or the error to show in the panel.</summary>
    private async Task<AppError?> SendAsync(string path, HttpMethod method)
    {
        try
        {
            var response = await RequestAsync(path, method);
            if (response.Status == 401)
            {
                auth.InvalidateAccessToken();
                return Classify(await RequestAsync(path, method));
            }
            return Classify(response);
        }
        catch (AppErrorException e)
        {
            return e.Error;
        }
        catch (Exception e)
        {
            return new AppError.Network(e.Message);
        }
    }

    private async Task<HttpResponse> RequestAsync(string path, HttpMethod method)
    {
        using var request = new HttpRequestMessage(method, Base + path);
        request.Headers.Add("Authorization", $"Bearer {await auth.AccessTokenAsync()}");
        // Spotify rejects a PUT with no body and no length header.
        request.Content = new ByteArrayContent([]);
        return await http.SendAsync(request);
    }

    private static AppError? Classify(HttpResponse response) => response.Status switch
    {
        >= 200 and < 300 => null,
        401 => new AppError.SessionExpired(),
        // 403 covers both "free account" and "token predates the control
        // scope"; only the body tells them apart.
        403 => response.Text.Contains("scope", StringComparison.OrdinalIgnoreCase)
            ? new AppError.ControlScopeMissing()
            : new AppError.PremiumRequired(),
        404 => new AppError.NoActiveDevice(),
        429 => new AppError.RateLimited(
            response.Headers.TryGetValue("Retry-After", out var value)
                && double.TryParse(value, out var seconds) ? seconds : 5),
        var status => new AppError.BadResponse(status),
    };
}
